crypto/fipsmodule/service_indicator/service_indicator_test.cc

```diff
@@ -63,6 +67,242 @@ static const uint8_t kAESCBCCiphertext[64] = {
+  // func is the hash function for HMAC to test.
+  const EVP_MD *(*func)(void);
+  // input is a NUL-terminated string to hash.
+  const uint8_t *input;
+  // expected_digest is the expected digest.
+  const uint8_t *expected_digest;
+  // expected to be approved or not.
+  const int expect_approved;
+} kHMACTestVectors[] = {
+    { EVP_sha1, kPlaintext, kHMACOutput_sha1, AWSLC_APPROVED },
+    { EVP_sha224, kPlaintext, kHMACOutput_sha224, AWSLC_APPROVED },
+    { EVP_sha256, kPlaintext, kHMACOutput_sha256, AWSLC_APPROVED },
+    { EVP_sha384, kPlaintext, kHMACOutput_sha384, AWSLC_APPROVED },
+    { EVP_sha512, kPlaintext, kHMACOutput_sha512, AWSLC_APPROVED },
+    { EVP_sha512_256, kPlaintext, nullptr, AWSLC_NOT_APPROVED }
+};
+
+class HMAC_ServiceIndicatorTest : public testing::TestWithParam<HMACTestVector> {};
+
+INSTANTIATE_TEST_SUITE_P(All, HMAC_ServiceIndicatorTest, testing::ValuesIn(kHMACTestVectors));
+
+TEST_P(HMAC_ServiceIndicatorTest, HMACTest) {
+  const HMACTestVector &hmacTestVector = GetParam();
+
+  int approved = AWSLC_NOT_APPROVED;
+  const uint8_t kHMACKey[64] = {0};
+  std::string digest_str;
+  const EVP_MD *digest = hmacTestVector.func();
+  std::vector<uint8_t> key(kHMACKey, kHMACKey + sizeof(kHMACKey));
+  std::vector<uint8_t> input(hmacTestVector.input, hmacTestVector.input + sizeof(hmacTestVector.input));
+
+  unsigned expected_mac_len = EVP_MD_size(digest);
+  std::unique_ptr<uint8_t[]> mac(new uint8_t[expected_mac_len]);
+  unsigned mac_len;
+
+  // Test running the HMAC interfaces one by one directly, and check
+  // |HMAC_Final| for approval at the end.
+  bssl::ScopedHMAC_CTX ctx;
+  ASSERT_TRUE(HMAC_Init_ex(ctx.get(), key.data(), key.size(), digest, nullptr));
+  ASSERT_TRUE(HMAC_Update(ctx.get(), hmacTestVector.input, input.size()));
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(HMAC_Final(ctx.get(), mac.get(), &mac_len)));
+  ASSERT_EQ(approved, hmacTestVector.expect_approved);
+  if (hmacTestVector.expect_approved) {
+    ASSERT_TRUE(check_test(hmacTestVector.expected_digest, mac.get(), mac_len, "HMAC KAT"));
+  }
+
+  // Test using the one-shot API for approval.
+  CALL_SERVICE_AND_CHECK_APPROVED(approved, ASSERT_TRUE(HMAC(digest, key.data(), key.size(), input.data(), input.size(),
+                                                             mac.get(), &mac_len)));
+  ASSERT_EQ(approved, hmacTestVector.expect_approved);
+  if (hmacTestVector.expect_approved) {
+    ASSERT_TRUE(check_test(hmacTestVector.expected_digest, mac.get(), mac_len, "HMAC KAT"));
+  }
+}
+
+TEST(ServiceIndicatorTest, CMAC) {
+  int approved = AWSLC_NOT_APPROVED;
+
+  uint8_t mac[16];
+  size_t out_len;
+  bssl::UniquePtr<CMAC_CTX> ctx(CMAC_CTX_new());
+  ASSERT_TRUE(ctx);
+
+  // Test running the CMAC interfaces one by one directly, and check
+  // |CMAC_Final| for approval at the end.
+  ASSERT_TRUE(CMAC_Init(ctx.get(), kAESKey, sizeof(kAESKey), EVP_aes_128_cbc(), nullptr));
+  ASSERT_TRUE(CMAC_Reset(ctx.get()));
```

Comment:
Do you need to CMAC_Reset when you just did CMAC_Init, or it's just for the sake of including it in the test?